HW1/P5/P5_bfs.py

```diff
@@ -0,0 +1,53 @@
+from pyspark import SparkContext
+
+def split_link(text):
+    colon_split = text.split(": ")
+    return (int(colon_split[0]), map(int, colon_split[1].split(" ")))
+
+def title_index(name1, name2, rdd):
+    return (rdd.filter(lambda x: x[1] == name1).first()[0], rdd.filter(lambda x: x[1] == name2).first()[0])
+
+def add_to_path(node, acc, index1, index2):
+    new_paths = []
+    last_node, pair = node
+    neighbors, curr_path = pair
+    neighbors = neighbors[0]
+    for neighbor in neighbors:
+        new_paths.append((neighbor, curr_path + [neighbor]))
+        if index1==curr_path[0] and index2 == neighbor:
+            acc.add(1)
+    return new_paths
+
+def shortest_path(index1, index2, links_rdd, numPartitions = 64):
+    solutions_found = links_rdd.context.accumulator(0)
+    nodes_to_search = links_rdd.filter(lambda x: x[0] == index1).partitionBy(numPartitions)
+    path_rdd = nodes_to_search.map(lambda x: (x[0], [x[0]])).partitionBy(numPartitions)
+    solution_rdd = path_rdd
+    while solutions_found.value == 0:
+        join_rdd = links_rdd.join(nodes_to_search).join(path_rdd)
+        path_rdd = join_rdd.flatMap(lambda x: add_to_path(x, solutions_found, index1, index2)).partitionBy(numPartitions)
+        nodes_to_search = path_rdd.map(lambda x: x[0]).distinct().map(lambda x: (x, [])).partitionBy(numPartitions)
+        solution_rdd = path_rdd.filter(lambda x: x[0] == index2 and x[1][0] == index1).map(lambda x: x[1]).collect()
```

Comment:
It's strange, for problem 4 I cached the main rdd in the while loop and was able to shave a second off the computation for captain america (from ~4 to ~3 seconds). I tried the same here with path_rdd and the computation actually took 30 seconds LONGER (~260 to ~290 seconds). Any ideas why that is?